panther_power_control/src/power_board_node.py

```diff
@@ -171,54 +179,62 @@ def _motor_controllers_state_cb(self, msg: DriverState) -> None:
-            if pin == self._pins.E_STOP_RESET:
-                self._e_stop_interrupt_time = rospy.get_time()
-
     def _publish_pin_state_cb(self, *args) -> None:
+        pin_state_time = rospy.get_time()
         with self._pins_lock:
-            charger_state = self._read_pin(self._pins.CHRG_SENSE)
-            self._publish_io_state('charger_connected', charger_state)
+            charger_state = not self._lines['CHRG_SENSE'].get_value()
+            shdn_init_val = self._lines['SHDN_INIT'].get_value()
+            estop_state = not self._lines['E_STOP_RESET'].get_value()
+
+        self._publish_io_state('charger_connected', charger_state)
+
+        # filter short spikes of voltage on GPIO
+        if math.isinf(self._shdn_init_detect_time) and shdn_init_val:
+            self._shdn_init_detect_time = pin_state_time
+        elif pin_state_time - self._shdn_init_detect_time > self._gpio_wait:
+            if shdn_init_val and not self._last_shdn_init_state:
+                self._last_shdn_init_state = shdn_init_val
+                rospy.loginfo(f'[{rospy.get_name()}] Shutdown button pressed.')
+                self._publish_io_state('power_button', True)
+            self._shdn_init_detect_time = float('inf')
 
-            # filter short spikes of voltage on GPIO
-            if rospy.get_time() - self._chrg_sense_interrupt_time > self._gpio_wait:
-                if self._read_pin(self._pins.SHDN_INIT):
-                    rospy.loginfo(f'[{rospy.get_name()}] Shutdown button pressed.')
-                    self._publish_io_state('power_button', True)
-                self._chrg_sense_interrupt_time = float('inf')
+        with self._e_stop_lock:
+            last_e_stop_state = self._last_e_stop_state
 
-            if rospy.get_time() - self._e_stop_interrupt_time > self._gpio_wait:
-                self._e_stop_event()
-                self._e_stop_interrupt_time = float('inf')
+        if math.isinf(self._e_stop_pressed_time) and estop_state != last_e_stop_state:
+            self._e_stop_pressed_time = pin_state_time
+        elif pin_state_time - self._e_stop_pressed_time > self._gpio_wait:
+            self._e_stop_event()
+            self._e_stop_pressed_time = float('inf')
 
     def _watchdog_cb(self, *args) -> None:
-        self._watchdog()
+        with self._watchdog_lock:
+            self._watchdog()
 
     def _aux_power_enable_cb(self, req: SetBoolRequest) -> SetBoolResponse:
-        res = self._set_bool_srv_handle(req.data, self._pins.AUX_PW_EN, 'Aux power enable')
+        res = self._set_bool_srv_handle(req.data, 'AUX_PW_EN', 'Aux power enable')
         if res.success:
             self._publish_io_state('aux_power', req.data)
         return res
 
     def _charger_enable_cb(self, req: SetBoolRequest) -> SetBoolResponse:
-        res = self._set_bool_srv_handle(not req.data, self._pins.CHRG_DISABLE, 'Charger disable')
+        res = self._set_bool_srv_handle(not req.data, 'CHRG_DISABLE', 'Charger disable')
```

Comment:
this is up for discussion, but I would prefer having type instead of string (`PatherGPIO` but with strings instead of numbers), because it's harder to make a typo. If we have a type, where values `pins.CHRG_DISABLE` are mapped to string `CHRG_DISABLE` and use `pins.CHRG_DISABLE` it will be harder to make a mistake.

Reply:
ok to leave with strings, using wrong name will cause error